lighthouse-core/computed/layout-shift-variants.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const makeComputedArtifact = require('./computed-artifact.js');
+const TraceOfTab = require('./trace-of-tab.js');
+
+/** @typedef {{ts: number, score: number}} LayoutShiftEvent */
+
+/**
+ * @fileoverview An implementation of the Layout Shift variants being considered
+ * as successors to CLS. This file will be removed when a variant is chosen and
+ * promoted to a Core Web Vital.
+ * @see https://web.dev/better-layout-shift-metric/
+ * @see https://github.com/mmocny/web-vitals/wiki/Snippets-for-LSN-using-PerformanceObserver
+ */
+
+class LayoutShiftVariants {
+  /**
+   * Calculates the average LayoutShift score across clusters of LayoutShift
+   * events, where a new cluster is created when there's a gap of at least 5s
+   * since the last LayoutShift event.
+   * @param {Array<LayoutShiftEvent>} layoutShiftEvents
+   * @return {number}
+   */
+  static avgSessionGap5s(layoutShiftEvents) {
+    if (layoutShiftEvents.length === 0) return 0;
+
+    let score = 0;
+    let clusterCount = 0;
+    let prevTs = Number.NEGATIVE_INFINITY;
+
+    for (const event of layoutShiftEvents) {
+      if (event.ts - prevTs > 5_000_000) clusterCount++;
+      prevTs = event.ts;
+      score += event.score;
```

Comment:
are we ignoring framehole here for now?

Reply:
> are we ignoring framehole here for now?

yes, see the `// Main frame only for now` comment below. Since `cumulativeLayoutShiftAllFrames` is currently in the data gathering stage it seemed like it made the most sense to make the direct comparison to main-frame-only CLS. We could add all frame versions of these variants if it made sense, though?

Reply:
I'd like to include the allFrames variants as part of this yeah.

Reply:
> I'd like to include the allFrames variants as part of this yeah.

With the size this PR has grown, would you be ok with a follow-up?

(With #12023 it sounds like we can't do it right until m90, which won't be used by HTTP archive until April (?), so we could maybe wait for @adamraine's refactor as well if that's incoming)

Reply:
> > I'd like to include the allFrames variants as part of this yeah.
> 
> With the size this PR has grown, would you be ok with a follow-up?

sure